dts/core-2022: Add DT Overlay [REVPI-2271]

zhan-kunbus opened the pull request:
Signed-off-by: Zhi Han <z.han@kunbus.com>

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1.1-4s-overlay.dts:
```diff
@@ -0,0 +1,65 @@
+		__overlay__ {
+			/delete-node/ eth1_2_reset_pins;
+
+			pileft_int_pins: pileft_int_pins {
```
```suggestion
			pileft_pins: pileft_pins {
```

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1.1-4s-overlay.dts:
```diff
@@ -0,0 +1,65 @@
+			/delete-node/ eth1_2_reset_pins;
+
+			pileft_int_pins: pileft_int_pins {
+				brcm,pins     = <13>;
```
Add the reset gpio as well!

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1.1-4s-overlay.dts:
```diff
@@ -0,0 +1,65 @@
+				brcm,pull     = <BCM2835_PUD_OFF>;
+			};
+
+			piright_int_pins: piright_int_pins {
```
```suggestion
			piright_pins: piright_pins {
```

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1.1-4s-overlay.dts:
```diff
@@ -0,0 +1,65 @@
+			};
+
+			piright_int_pins: piright_int_pins {
+				brcm,pins     = <12>;
```
Also add the reset gpio!

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1.1-4s-overlay.dts:
```diff
@@ -0,0 +1,65 @@
+		__overlay__ {
+			pileft: ethernet@0 {
+				pinctrl-names = "default";
+				pinctrl-0 = <&pileft_int_pins>;
```
```suggestion
				pinctrl-0 = <&pileft_pins>;
```

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1.1-4s-overlay.dts:
```diff
@@ -0,0 +1,65 @@
+				interrupts = <12 IRQ_TYPE_LEVEL_LOW>;
+
+				reset-gpios = <&gpio 20 GPIO_ACTIVE_HIGH>;
+				vdd-supply = <>;
```
```suggestion
```

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1.1-4s-overlay.dts:
```diff
@@ -0,0 +1,65 @@
+				interrupts = <13 IRQ_TYPE_LEVEL_LOW>;
+
+				reset-gpios = <&gpio 21 GPIO_ACTIVE_HIGH>;
+				vdd-supply = <>;
```
```suggestion
```

zhan-kunbus replied:
without this, the label  pb_rst will be referred

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1.1-4s-overlay.dts:
```diff
@@ -0,0 +1,65 @@
+
+			piright: ethernet@1 {
+				pinctrl-names = "default";
+				pinctrl-0 = <&piright_int_pins>;
```
```suggestion
				pinctrl-0 = <&piright_pins>;
```

iluminat23 commented:
With the Core 1.1 the Pull-Up for the i2c have been changed. We expect better results and thus should be able to run the i2c with a higher clock rate. Please check which clock rates are supported by the devices connected to the i2c and the host controller and set the i2c clock rate to an appropriate value.
Also we have now added buffers to the sclk and the mosi lines. Which should make the signal better and thus enable us to use fater clock rates. Please check which clock rates are supported by the devices and the host controller and set the clock rates for the device to an appropriate value.

zhan-kunbus commented:
> With the Core 1.1 the Pull-Up for the i2c have been changed. We expect better results and thus should be able to run the i2c with a higher clock rate. Please check which clock rates are supported by the devices connected to the i2c and the host controller and set the i2c clock rate to an appropriate value. Also we have now added buffers to the sclk and the mosi lines. Which should make the signal better and thus enable us to use fater clock rates. Please check which clock rates are supported by the devices and the host controller and set the clock rates for the device to an appropriate value.

Is this the right way to set the clock/baudrate here: https://www.raspberrypi-spy.co.uk/2018/02/change-raspberry-pi-i2c-bus-speed/

iluminat23 commented:
For the future, please use a correct branch name!

iluminat23 commented:
> Is this the right way to set the clock/baudrate here: https://www.raspberrypi-spy.co.uk/2018/02/change-raspberry-pi-i2c-bus-speed/

https://github.com/RevolutionPi/linux/blob/e97f07c97ee61f42a21bf8b2a4acc6ed589feaac/Documentation/devicetree/bindings/i2c/i2c.txt#L26
https://github.com/RevolutionPi/linux/blob/e97f07c97ee61f42a21bf8b2a4acc6ed589feaac/Documentation/devicetree/bindings/i2c/brcm%2Cbcm2835-i2c.txt#L21

zhan-kunbus commented:
> With the Core 1.1 the Pull-Up for the i2c have been changed. We expect better results and thus should be able to run the i2c with a higher clock rate. Please check which clock rates are supported by the devices connected to the i2c and the host controller and set the i2c clock rate to an appropriate value. Also we have now added buffers to the sclk and the mosi lines. Which should make the signal better and thus enable us to use fater clock rates. Please check which clock rates are supported by the devices and the host controller and set the clock rates for the device to an appropriate value.

The crypto chip supports frequency of max.1MHz (Features section in DS), and RTC chip supports max.400kHz (section 13.2 in DS), so it looks that the 400kHz can be idealist used. However, testing on the HW is always needed.

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1.1-overlay.dts:
```diff
@@ -0,0 +1,71 @@
+	fragment@8 {
+		target = <&spi0>;
+		__overlay__ {
+			pileft: ethernet@0 {
```
The SPI device are missing max speed configuration.
https://github.com/RevolutionPi/linux/blob/e97f07c97ee61f42a21bf8b2a4acc6ed589feaac/arch/arm/boot/dts/overlays/revpi-core-overlay.dts#L204
Please check the max SPI freq. the device can handle.

zhan-kunbus replied:
the spi-max-frequency is inherited from core-overlay

iluminat23 replied:
We now have buffers in the clock and MOSI lines, such the freqency could be much faster.

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1.1-overlay.dts:
```diff
@@ -0,0 +1,71 @@
+	fragment@3 {
+		target = <&i2c1>;
+		__overlay__ {
+			/* I2C fast mode */
```
Please add a note why 400kHz are set here. For the next developer it might seem totally random.

zhan-kunbus replied:
Just tested it on Core 1.1,  any increase of the clock-frequency will cause crypto chip acessing fail. (rtc works always)
`pi@RevPi1234:~ $ piSerial
WARNING: Can't write to `/dev/i2c-1': Remote I/O error
WARNING: Can't write to `/dev/i2c-1': Remote I/O error
WARNING: Can't write to `/dev/i2c-1': Remote I/O error
WARNING: Can't write to `/dev/i2c-1': Remote I/O error
WARNING: Can't write to `/dev/i2c-1': Remote I/O error
ERROR: FATAL: Can't get serial number from device
`

zhan-kunbus replied:
I will remove this overlay node.

iluminat23 reviewed (changes requested):
Have tried to run the overlay on a Core? I think i should start, with some errors about the ksz8851 interrupts. But you can check the fdt. If everything is as intended.
Have you decompiled the .dtbo and checked everything works as inteded?

iluminat23 commented:
From the commit message:
> As the core 1.1 is based on the old core (LP0124R02)

This is wrong, the Core 1.1 is based on the LP0124R06. But the LP is not the relevant part. The BG is relevant. => BG0137R07. The LP is the layout but doesn't say anything about the soldered components. A new BG might be created just by removing parts from the same LP.

zhan-kunbus commented:
> 

maybe say it as :  "As the core 1.1 is based on the old core (BG0137R03)" ?

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1.1-overlay.dts:
```diff
@@ -0,0 +1,75 @@
+		__overlay__ {
+			/*
+			 * The frequency is 400kHZ when I2C in fast mode.
+			 * With 2K Pullup resistor, it is expected.
```
This comment is not helpfull.
It would be good to give a hint the the crypto chip is able to do 1MHz and the RTC only 400kHz. Also add that the core 1.1 has different *pull-ups as the original core and higher frequencies are now possible.

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1.1-overlay.dts:
```diff
@@ -0,0 +1,75 @@
+	fragment@8 {
+		target = <&spi0>;
+		__overlay__ {
+			pileft: ethernet@0 {
```
You could delete the nodes and recreate them?

zhan-kunbus commented:
> Have tried to run the overlay on a Core? I think i should start, with some errors about the ksz8851 interrupts. But you can check the fdt. If everything is as intended. Have you decompiled the .dtbo and checked everything works as inteded?

Tried it on Core with CM4S, with command dtoverlay it works basically as expected, but not with bootloader.   As I tested, change the filename of the dtbo as no dot in it, then it works as expected. 

The dtc -O dts <dtbo file> gives no output with dtc version 1.4.7, seems the reference of &gpio failed

`<stdout>: Warning (gpios_property): /__fixups__:gpio: property size (737) is invalid, expected multiple of 4
dtc: livetree.c:565: get_node_by_phandle: Assertion generate_fixups failed.
Aborted
`
However, when I try it with dtc version > 1.5.0, the it give the complete output even the warning of missing reference of &gpio are still there

`<stdout>: Warning (interrupts_property): /fragment@8/__overlay__/ethernet@0:interrupt-parent: Invalid phandle
<stdout>: Warning (interrupts_property): /fragment@8/__overlay__/ethernet@0: Missing interrupt-parent
<stdout>: Warning (interrupts_property): /fragment@8/__overlay__/ethernet@1:interrupt-parent: Invalid phandle
<stdout>: Warning (interrupts_property): /fragment@8/__overlay__/ethernet@1: Missing interrupt-parent
/dts-v1/;
`

iluminat23 commented:
This doesn't look so good.
IS the resultig fdt ok? Are the phandle references for the ethernet nodes correct?

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1.1-overlay.dts:
```diff
@@ -0,0 +1,75 @@
+			pileft: ethernet@0 {
+				pinctrl-names = "default";
+				pinctrl-0 = <&pileft_pins>;
+				/*interrupt-parent = <&gpio>;*/
```
Is this commented out by intention?

zhan-kunbus commented:
> This doesn't look so good. IS the resultig fdt ok? Are the phandle references for the ethernet nodes correct?

the fdt works fine with it.
`002473.871: brfs: File read: /mfs/sd/config.txt
002474.496: dtparam: audio=on
002485.672: brfs: File read: 1841 bytes
002496.970: brfs: File read: /mfs/sd/overlays/revpi-core-1-1.dtbo
002542.475: Loaded overlay 'revpi-core-1-1'
002542.515: dtparam: eth0_mac_hi=0xc83ea7fe
002545.640: dtparam: eth0_mac_lo=0x1234
002743.021: brfs: File read: 6075 bytes
`
ethernet interfaces works:
`pi@RevPi1234:~ $ dmesg | grep ks8851
[    8.522086] ks8851 spi0.1: supply vdd-io not found, using dummy regulator
[    8.524110] ks8851 spi0.1: supply vdd not found, using dummy regulator
[    8.546364] ks8851 spi0.1: message enable is 0
[    8.549629] ks8851 spi0.1 eth1: revision 1, MAC 46:81:0f:dd:53:09, IRQ 199, no EEPROM
[    8.550373] ks8851 spi0.0: supply vdd-io not found, using dummy regulator
[    8.550607] ks8851 spi0.0: supply vdd not found, using dummy regulator
[    8.572840] ks8851 spi0.0: message enable is 0
[    8.575927] ks8851 spi0.0 eth2: revision 1, MAC 0e:da:62:d8:10:a6, IRQ 200, no EEPROM
[   17.998921] ks8851 spi0.1 piright: renamed from eth1
[   18.092680] ks8851 spi0.0 pileft: renamed from eth2
`

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1-1-overlay.dts:
```diff
@@ -0,0 +1,64 @@
+				#interrupt-cells = <2>;
+				interrupts = <13 IRQ_TYPE_LEVEL_LOW>;
+				reset-gpios = <&gpio 21 GPIO_ACTIVE_HIGH>;
+				vdd-supply = <>;
```
The empty vdd-supply looks ugly. Wouldn't it be better to delete the whole node and recreate it?
@zhan-kunbus

zhan-kunbus replied:
I think yes, just following info would be repeated

`compatible = "micrel,ks8851";
reg = <0>;
spi-max-frequency = <20000000>;	/* <= 40 MHz */
interrupts = <0x80000000>;  /* not connected */
status = "okay";`

zhan-kunbus commented on arch/arm/boot/dts/overlays/revpi-core-1-1-overlay.dts:
```diff
@@ -0,0 +1,64 @@
+	fragment@0 {
+		target-path = "/";
+		__overlay__ {
+			compatible = "kunbus,revpi-core-1.1", "brcm,bcm2711";
```
kunbus,revpi-core-1.1 or kunbus,revpi-core-1-1 ? @iluminat23

zhan-kunbus replied:
we shouldn't use period, as spec:
_"The compatible string should consist only of lowercase letters, digits and dashes, and should start with a letter."_

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1-1-overlay.dts:
```diff
@@ -0,0 +1,64 @@
+			};
+		};
+	};
+};
```
eth0_mac_hi and eth0_mac_left are not used any more. We have an eeprom attached to the USB phys.

zhan-kunbus replied:
I assume you mean the eth0_mac_* is not needed any more.   

And how about pileft_mac_* and piright_mac_* ? Can the be obtained from hat-eeprom or some where?  But anyway, it seems they are never gotten used, am I right?

zhan-kunbus replied:
I deleted the eth0_mac_*

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1-1-overlay.dts:
```diff
@@ -0,0 +1,85 @@
+		};
+	};
+
+/* The spi-max-frequency is set to 20MHz because of hardware limitations.
+The KSZ8851 support up to 40MHz, but because of the poor signal quality of the
+SPI on the BCM28xx/BCM27xx only ~29MHz can be achieved. To be on the safe side
+we limit the spi-max-frequency to 20MHz.
+Even for this the drive strength needs to be at least 8mA. */
```
```suggestion
/*
   The spi-max-frequency is set to 20MHz because of hardware limitations.
   The KSZ8851 support up to 40MHz, but because of the poor signal quality of the
   SPI on the BCM28xx/BCM27xx only ~29MHz can be achieved. To be on the safe
   side we limit the spi-max-frequency to 20MHz.
   Even for this the drive strength needs to be at least 8mA.
*/
```
The issue with the comment at this location is the spi-max-freq is not set here. So it doesn't make any sense. Could you please add this comment to the revpi-core-overlay? And in this commit message give a hint that this is still relevant for the revpi-core-2022.

zhan-kunbus replied:
but this problem is not exist on old Core, isn't it?

iluminat23 replied:
This problem existed always and is also not fixed with the changes in the current core redesign

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1-1-overlay.dts:
```diff
@@ -0,0 +1,85 @@
+/*
+ * Device tree overlay for Revolution Pi by KUNBUS
+ *
+ * RevPi Core 1.1
```
```suggestion
 * RevPi Core 3 (2022)
```

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1-1-overlay.dts:
```diff
@@ -0,0 +1,85 @@
+	fragment@0 {
+		target-path = "/";
+		__overlay__ {
+			compatible = "kunbus,revpi-core-1-1", "brcm,bcm2711";
```
```suggestion
			compatible = "kunbus,revpi-core-2022", "brcm,bcm2837";
```

zhan-kunbus replied:
compatible = "kunbus,revpi-core-2022", "brcm,bcm2711";

iluminat23 replied:
No, this is Core 3 not Core S

iluminat23 commented on arch/arm/boot/dts/overlays/Makefile:
```diff
@@ -175,6 +175,7 @@ dtbo-$(CONFIG_ARCH_BCM2835) += \
 	revpi-flat-dt-blob.dtbo \
 	revpi-core.dtbo \
 	revpi-core-dt-blob.dtbo \
+	revpi-core-1-1.dtbo \
```
```suggestion
	revpi-core-2022.dtbo \
```

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-1-1-overlay.dts:
```diff
@@ -0,0 +1,85 @@
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/{
+	compatible = "brcm,bcm2711", "brcm,bcm2837";
```
```suggestion
	compatible = "brcm,bcm2837";
```

zhan-kunbus replied:
should it not be "brcm,bcm2711"

iluminat23 replied:
No, as this is Core 3 not Core S

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-overlay.dts:
```diff
@@ -197,6 +197,14 @@
+			 SPI on the BCM28xx/BCM27xx only ~29MHz can be achieved. To be on the safe
+			 side we limit the spi-max-frequency to 20MHz.
+			 Even for this the drive strength needs to be at least 8mA.
+			 This is still relevant for the revpi-core-2022.
```
```suggestion
```
Add this to the commit message which adds the core-2022 overlay.

iluminat23 commented:
The commit message of the overlay still includes the name core1.1.

iluminat23 commented:
It doesn't make much sense to say the Core is based on the LP0124R02. The LP is only the circuit board (**L**eiter**p**latte) without any components. For example the Core S and the Core SE use the same LP but use different components.
We use the term **B**au**g**ruppe (BG, assembly) for the whole board with components. The Core 3 (2022) uses the BG0137R07 the "old" Core 3 uses the BG0137R02.
Why don't you just say that the new core is based on the old core. And then describe the relevant changes.

iluminat23 commented on arch/arm/boot/dts/overlays/revpi-core-2022-overlay.dts:
```diff
@@ -0,0 +1,80 @@
+/*
+ * Device tree overlay for Revolution Pi by KUNBUS
+ *
+ * RevPi Core 3 (2022)
```
```suggestion
 * RevPi Core 3 (2022)
 * RevPi Core 3+ (2022)
```
Sorry for the late comment, I have only thought about this, this morning.

iluminat23 commented:
As you are adding a new overlay the summery prefix should be dts: not dts/core-2020.
The summary should then include the nae of the overlay or device:
so something like this: dts: Add RevPi Core 3/3+ (2022) overlay

iluminat23 reviewed (changes requested):
```
The new core is based on the old core (revpi-core-overlay.dts) and
some enhancements are made,
```
This is some what useless. Obviously we made some enhancements. And writing the new core is based on the old core is also a bit strange.
You could write something like this:
This overlay is based on the original RevPi Core overlay. As the RevPi Core (2022) introduces some changes to the hardware the following changes have been made to the overlay.
```
 * regulator and pins for eth chips resetting are removed
 * new pins for the eth chips are added and get applied
 ```
Both network phys on the PiBridge now have there own reset line. Also the polarity of the resets are now as expected by the driver. And the delay circuit has also been removed. We can now remove the the pbrst regulator.
Also the interrupt line of the phys are now connected to gpios on the compute module.
```
* overrides for mac of eth0 are removed as hat-eeprom get applied
```
The USB network interfaces has its own eeprom to store the MAC address. Thus we don't need to set the MAC address from this overlay any more.

@zhan-kunbus FYI: The MAC address is stored in an eeprom attached to the USB phy to in the hat-eeprom. The hat eeprom contains the mac address as well, but this is just a precaution, if we might have a network device which can't be configured through an eeprom.
